packages/console/src/hooks/useFileDecryption.ts

```diff
@@ -0,0 +1,153 @@
+    const url = URL.createObjectURL(blob)
+    const a = document.createElement('a')
+    a.href = url
+    a.download = filename
+    document.body.appendChild(a)
+    a.click()
+    document.body.removeChild(a)
+    URL.revokeObjectURL(url)
+  }
+
+  const streamToBlob = async (stream: ReadableStream, mimeType?: string): Promise<Blob> => {
+    const reader = stream.getReader();
+    const chunks: Uint8Array[] = [];
+    let done = false;
+    let totalBytes = 0;
+
+    while (!done) {
+      const { value, done: isDone } = await reader.read();
+      done = isDone;
+      if (value) {
+        chunks.push(value);
+        totalBytes += value.length;
+      }
+    }
+
+    return new Blob(chunks, { type: mimeType || 'application/octet-stream' });
+  };
+
+  const decryptAndDownload = async (cid: UnknownLink | string, filename?: string) => {
+    if (!client || !space || space.access?.type !== 'private') {
+      throw new Error('Invalid state: client, space, or private space access required')
+    }
+
+    setState({ loading: true, error: null, fileMetadata: undefined })
+
+    try {
+      // Create crypto adapter using shared KMS config
+      const cryptoAdapter = await createKMSAdapter()
+      if (!cryptoAdapter) {
+        throw new Error('KMS configuration required for decryption')
+      }
+
+      // Create encrypted client
+      const encryptedClient = await createEncryptedClient({
+        storachaClient: client,
+        cryptoAdapter
+      })
+
+      // Parse CID if it's a string
+      const encryptionMetadataCID = typeof cid === 'string' ? parseLink(cid) : cid
+      const proofs = client.proofs([
+        {
+          can: 'space/content/decrypt',
+          with: space.did()
+        }
+      ])
+      .map(proof => /* @type {import('@ucanto/interface').Delegation} */ (proof))
+      .filter(delegation => !delegation.capabilities.some(cap => cap.can === '*' || 
+          cap.can === 'ucan/attest' ||
+          cap.with === 'ucan:*'))
+      
+      const decryptDelegation = await decrypt.delegate({
+        issuer: client.agent.issuer,
+        audience: client.agent.issuer,
+        with: space.did(),
+        nb: {
+          resource: encryptionMetadataCID,
+        },
+        expiration: Math.floor(Date.now() / 1000) + 60 * 15, // 15 minutes
+        proofs,
+      })
+
+      // Downloads the encrypted file, and decrypts it locally
+      const { stream: decryptedStream, fileMetadata } = await encryptedClient.retrieveAndDecryptFile(
+        encryptionMetadataCID,
+        {
+          spaceDID: space.did(),
+          decryptDelegation, 
+          proofs,
+        }
+      )
+
+      // Use metadata for filename and MIME type, with robust fallbacks
+      let finalFilename: string
+      if (fileMetadata?.name) {
+        finalFilename = fileMetadata.name
+      } else if (fileMetadata?.extension) {
+        finalFilename = `decrypted-file-${Date.now()}.${fileMetadata.extension}`
```

Comment:
curious how often these other cases will actually get hit, if we now always save the filename in the data

Reply:
That was a fallback that I implemented to be able to decrypt the initial files I had uploaded when there was no file metadata handling. I don't think we will hit that in production.